### PR TITLE
feat: t3-env

### PR DESCRIPTION
**Pull request by tristancamejo**
Implements https://github.com/ixahmedxi/noodle/issues/245

### CODE REVIEW

**Review by ajitzero — APPROVED** [review]

File: packages/env/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  server: {
+    DATABASE_URL: z.string().url(),
+    REDIS_URL: z.string().url(),
+    REDIS_TOKEN: z.string().min(1),
+    VERCEL_URL: z.string().url().optional(),
+    PORT: z.string().optional(),
+    GITHUB_CLIENT_ID: z.string().min(1),
+    GITHUB_CLIENT_SECRET: z.string().min(1),
+    NODE_ENV: z.enum(['development', 'test', 'production']),
+  },
```
Comment (ajitzero) [inline packages/env/src/index.ts:19]: `VERCEL_URL` & `PORT` are optional only if `typeof window` is `undefined`, i.e., on client side, which is never the case here.

In that case, we need either `VERCEL_URL` or `PORT` to be configured (in order). I'm basing this off the [`getBaseUrl`](https://github.com/ixahmedxi/noodle/blob/2bc1e682376f13c4bd5c23725831415b75eae3ff/apps/web/src/utils/getBaseUrl.ts#L1C14-L1C24) function. I suggest using zod's [`.refine`](https://zod.dev/?id=refine) for more type-safety.

Comment (ixahmedxi) [reply packages/env/src/index.ts:19]: if `typeof window` is `undefined` that means it's server side not client side

Comment (ixahmedxi) [reply packages/env/src/index.ts:19]: which means this is correct, as that it's defined in the server environment variables

Comment (ajitzero) [reply packages/env/src/index.ts:19]: Ah, yes, you're correct. I wanted to point out that we could use `.refine` to enforce requiring either `VERCEL_URL` or `PORT`. I overlooked that it's on the server side anyway.

**Comment by ixahmedxi** [comment]
Hey so this is gonna need some work with my preview environment deployments, nothing wrong with the code. I'm gonna be on it tonight and tomorrow.

**Pull request closed without merging**